torch/distributed/_symmetric_memory/_nvshmem_triton.py

```diff
@@ -358,60 +358,101 @@ def broadcastmem_block(team, dest, source, size_bytes, pe_root, _semantic=None):
-            },
-            is_pure=False,
-            _semantic=_semantic,
-        )
+    # Reduction Operation
+    @core.extern  # type: ignore[misc]
+    def reduce(team, dest, source, nreduce, operation: str, dtype_id, _semantic=None):  # type: ignore[no-untyped-def]
+        """
+        Performs a collective reduction operation on symmetric data across a team of PEs.
+
+        This function provides a generic interface to NVSHMEM reduction operations,
+        automatically selecting the appropriate NVSHMEM function based on the data type
+        and operation specified.
+        Args:
+            team (int64): The team handle (0 for NVSHMEM_TEAM_WORLD).
+            dest (pointer): Destination pointer where reduction results are stored.
+            source (pointer): Source pointer containing data to be reduced.
+            nreduce (int64): Number of elements to reduce.
+            operation (str): Reduction operation ("sum", "max", "min", "prod").
+            dtype_id: Data type specification - accepts torch.dtype, tl.dtype, str, or constexpr.
+            _semantic: Optional semantic information for Triton compilation.
+
+        Raises:
+            ValueError: If the operation is not supported.
+            TypeError: If the data type is not supported.
+
+        Example:
+            nvshmem.reduce(0, dest_ptr, src_ptr, 100, "sum", torch.float32)
+        """
+        # Mapping from PyTorch/Triton dtype names to NVSHMEM typenames
+        DTYPE_TO_NVSHMEM_MAP = {
+            "int8": "int8",
+            "int16": "int16",
+            "int32": "int32",
+            "int64": "int64",
+            "uint8": "uint8",
+            "uint16": "uint16",
+            "uint32": "uint32",
+            "uint64": "uint64",
+            "float16": "half",
+            "bfloat16": "bfloat16",
+            "float32": "float",
+            "float64": "double",
+        }
+
+        # Extract operation name from constexpr if needed
+        op_name = operation.value if hasattr(operation, "value") else operation
+
+        # Normalize dtype_id to a canonical string name
+        # Handle different input formats: tl.dtype, torch.dtype, str, constexpr[dtype]
+        if hasattr(dtype_id, "name"):
+            # Triton language dtype (e.g., tl.float32)
+            dtype_name = dtype_id.name
+        elif isinstance(dtype_id, str):
+            # Already a plain string name
+            dtype_name = dtype_id
+        elif hasattr(dtype_id, "value"):
+            # Constexpr wrapper around a dtype
+            inner_value = dtype_id.value
+            if hasattr(inner_value, "name"):
+                # Triton dtype inside constexpr
+                dtype_name = inner_value.name
+            else:
+                # PyTorch dtype inside constexpr
+                dtype_name = str(inner_value).replace("torch.", "")
+        else:
+            # PyTorch dtype (e.g., torch.float32)
+            dtype_name = str(dtype_id).replace("torch.", "")
```

Comment:
```suggestion
        inner_value = dtype_id.value if hasattr(dtype_id.value) else dtype_id 
        if hasattr(inner_value, "name"):
            # Triton dtype inside constexpr
            dtype_name = inner_value.name
        elif isinstance(inner_value, str):
            # Already a plain string name
            dtype_name = inner_value
        else:
            # PyTorch dtype (e.g., torch.float32)
            dtype_name = str(inner_value).replace("torch.", "")
```
Is there a more reliable way to distinguish between triton dtype and torch dtype, rather than relying on hasattr(..., "name")?

Reply:
In #159788 It's much simpler now because we only have triton dtypes coming in. we just have a map now!